File: session-02-dependency-injection/python/order/repository.py
from abc import ABC, abstractmethod

from order.model import Order


class OrderRepository(ABC):
    """Interface for order storage. Implementations are swapped via DI."""

    @abstractmethod
    def find_all(self) -> list[Order]:
        ...

    @abstractmethod
    def find_by_id(self, order_id: int) -> Order | None:
        ...


class InMemoryOrderRepository(OrderRepository):

    def __init__(self):
        self._orders = [
            Order(id=1, drink="Latte", status="ready"),
            Order(id=2, drink="Espresso", status="preparing"),
            Order(id=3, drink="Cappuccino", status="pending"),
        ]

    def find_all(self) -> list[Order]:
        return self._orders

    def find_by_id(self, order_id: int) -> Order | None:
        return next((o for o in self._orders if o.id == order_id), None)

File: session-02-dependency-injection/go/order/repository.go
package order

// OrderRepository is the interface. Any struct that implements these methods can be injected.
// Go interfaces are implicit: no "implements" keyword, no annotations, no decorators.
type OrderRepository interface {
	FindAll() []Order
	FindByID(id int) *Order
}

// InMemoryOrderRepository is the default implementation.
type InMemoryOrderRepository struct {
	orders []Order
}

func NewInMemoryOrderRepository() *InMemoryOrderRepository {
	return &InMemoryOrderRepository{
		orders: []Order{
			{ID: 1, Drink: "Latte", Status: "ready"},
			{ID: 2, Drink: "Espresso", Status: "preparing"},
			{ID: 3, Drink: "Cappuccino", Status: "pending"},
		},
	}
}

func (r *InMemoryOrderRepository) FindAll() []Order {
	return r.orders
}

func (r *InMemoryOrderRepository) FindByID(id int) *Order {
	for _, o := range r.orders {
		if o.ID == id {
			return &o
		}
	}
	return nil
}
